### PR TITLE
Adsorbate Placer

### PR DESCRIPTION
**Pull request by GabrielBram**
Basic functionality to place an adsorbate along a bond defined by the neighbours of the adsorption binding site. Based on the VSEPR model, assuming all species have one remaining lone pair. Rotation can then be performed using the bonding direction as the z-axis, the x-axis as the cross product of the z-axis and one other bond, and the y-axis as the cross product of the previously defined x and z axes. 

Minimum example shows how adsorption can be used to add simple one atom species to a given site, and more complex species may be appended and rotated into a desired position. 

(With other pull request in progress, will merge in the new upstream master once that is accepted).

### CODE REVIEW

File: carmm/build/adsorbate_placer.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+        atoms_site: Atoms object
+            Contains positions of the adsorption site.
+        bond_length: float
+            Length of the bond for adsorption.
+        ads_idx: integer
+            Index of the adsorbate atom bonding adsorbate and adsorption site.
+        site_idx: integer
+            Index of the site atom bonding adsorbate and adsorption site.
+        neighb_idx: integer
+            Index of neighbour used to form principle axes of rotation.
```
Comment (ikowalec) [inline carmm/build/adsorbate_placer.py:22]: Can we get this automatically using @JackWarren23's planes?

File: examples/build_adsorbate_place_and_rotate.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                                                           rotation=[45, 0, -45])
+
+    comp_pos1 = np.array([2.00311656e+01, 5.31509397e+00, 1.89702619e+00])
+    comp_pos2 = np.array([2.04405930e+01, 4.53669268e+00, 2.54385921e+00])
+
+    error_pos1 = np.linalg.norm(comp_pos1 - rotated_ads.positions[0], axis=-1)
+    error_pos2 = np.linalg.norm(comp_pos2 - rotated_ads.positions[-1], axis=-1)
+
+    assert (np.isclose(error_pos1, 0, rtol=0, atol=1e-06))
+    assert (np.isclose(error_pos2, 0, rtol=0, atol=1e-06))
```
Comment (ikowalec) [inline examples/build_adsorbate_place_and_rotate.py:33]: This currently returns an assertion error

Comment (ikowalec) [reply examples/build_adsorbate_place_and_rotate.py:33]: Just to confirm, I am using ase==3.22.1 and numpy==1.22.3, what about you?

Comment (GabrielBram) [reply examples/build_adsorbate_place_and_rotate.py]: Python==3.7.12
ASE==3.22.1
numpy==1.21.5 

No massive version differences. Pushed some basic printing commands into the assertion, so it should display just how badly we're off by. If it's say, 1e-05, the tolerance might be too tight. But anything larger than 1e-04 is cause for concern.

Comment (ikowalec) [reply examples/build_adsorbate_place_and_rotate.py]: The problem persists on my end, also in `view()` hydrogen index=3 and hydrogen index=9 are less than 0.5A apart, which would result in a failure in initialisation of FHI-aims calculation

Comment (ikowalec) [reply examples/build_adsorbate_place_and_rotate.py]: > Python==3.7.12 ASE==3.22.1 numpy==1.21.5
> 
> No massive version differences. Pushed some basic printing commands into the assertion, so it should display just how badly we're off by. If it's say, 1e-05, the tolerance might be too tight. But anything larger than 1e-04 is cause for concern.

I am on Python 3.10 on my local machine

Comment (GabrielBram) [reply examples/build_adsorbate_place_and_rotate.py]: Hmm. Problematic. What are the error values printed out?

Comment (ikowalec) [reply examples/build_adsorbate_place_and_rotate.py]: AssertionError: Error = 3.29883926784293

Comment (GabrielBram) [reply examples/build_adsorbate_place_and_rotate.py]: Aaaaah, now that is sizeable. Hmm. Let me make a new conda environment and test this out. Apologies about this.

Comment (ikowalec) [reply examples/build_adsorbate_place_and_rotate.py]: I see some dependency on `from carmm.analyse.neighbours import neighbours`, have you modified it at all?

Comment (GabrielBram) [reply examples/build_adsorbate_place_and_rotate.py]: I haven't modified any existing files to my knowledge. 

However, I rolled back to Python 3.1 and am getting the same error you observed - this doesn't happen with Python 3.7. Very curious. Will get back to you with the part of the code which is misbehaving.

Comment (GabrielBram) [reply examples/build_adsorbate_place_and_rotate.py]: For some reason beyond the ken of man - the order of the Neighbour positions lists changes between the two versions. So the code itself is working as intended, but reversing some rotations. Now put some stop gap statements to preserve order, but will tell you when everything is working.

Comment (GabrielBram) [reply examples/build_adsorbate_place_and_rotate.py]: Yeah, without ordering the neighbour list, python3.7 by default orders the neighbours [3,2], while 3.1 orders them [2,3]. A strange one, but the work around is simple.

**Comment by GabrielBram** [comment]
Bugs fixed - can this be merged in so I can merge in the other pull request?

**Pull request merged**